rclcpp/src/rclcpp/node_interfaces/node_graph.cpp

```diff
@@ -369,3 +369,87 @@ NodeGraph::count_graph_users()
 {
   return graph_users_count_.load();
 }
+
+static
+std::vector<rclcpp::TopicEndpointInfo>
+convert_to_topic_info_list(const rmw_topic_endpoint_info_array_t & info_array)
+{
+  std::vector<rclcpp::TopicEndpointInfo> topic_info_list;
+  for (size_t i = 0; i < info_array.count; ++i) {
+    topic_info_list.push_back(rclcpp::TopicEndpointInfo(info_array.info_array[i]));
+  }
+  return topic_info_list;
+}
+
+template <const char * EndpointType, typename FunctionT>
+static std::vector<rclcpp::TopicEndpointInfo>
+get_info_by_topic(
+  rclcpp::node_interfaces::NodeBaseInterface * node_base,
+  const std::string & topic_name,
+  bool no_mangle,
+  FunctionT rcl_get_info_by_topic)
+{
+  std::vector<rclcpp::TopicEndpointInfo> topic_info_list;
+
+  auto rcl_node_handle = node_base->get_rcl_node_handle();
+  auto fqdn = rclcpp::expand_topic_or_service_name(
```

Comment:
I'm curious, how does `expand_topic_or_service_name()` work without a `no_mangle` parameter?

Reply:
So, one thing is mangling, i.e. generate a valid middleware name (e.g. DDS) from the ROS name.
Another thing is expansion, i.e. get full names from relative names ('topic' -> '/namespace/topic/') and full names from private names ('~/topic' -> '/namespace/node_name/topic/').
Remapping should be always done after expansion, but it's missing here (can be added later, as it's a preexisting problem in `get_publisher_count` too).

We're everywhere always doing remapping and expansion, including when we do `no_mangle=true` (or when [this](https://github.com/ros2/rmw/blob/a3774531e9b7289574501275aafbc3f93120360e/rmw/include/rmw/types.h#L334) qos setting is set to `true`). IMO, always doing remapping and expansion is ok, though is arguable. Considering that, `expand_topic_or_service_name` doesn't need to know about `no_mangle`.

Reply:
Has this been resolved?

Reply:
I think we need to get a holistic summary of how we think `no_mangle` should be used with each of these functions, and then make sure the implementation matches. I feel like there have been several such conversations and potentially different solutions in each case. Also I haven't heard a compelling story for why one way or the other.

I would prefer to see clarity on this point before moving forward, even if that means we need to follow up after this pr and fix other functions.

Reply:
> Another thing is expansion, i.e. get full names from relative names ('topic' -> '/namespace/topic/') and full names from private names ('~/topic' -> '/namespace/node_name/topic/').

I meant to confirm if the expansion algorithm will work for a middleware topic names well? For example `rt/topic` -> `rt/namespace/topic`. (And I guess `rt/~/topic` is an invalid topic name regardless of the value of `no_mangle`.)

Reply:
I will answer to both this discussion and https://github.com/ros2/rclcpp/pull/960#discussion_r364215929.

> I think we need to get a holistic summary of how we think no_mangle should be used with each of these functions, and then make sure the implementation matches. I feel like there have been several such conversations and potentially different solutions in each case. Also I haven't heard a compelling story for why one way or the other.

Considering that the intention of `no_mangle=true` is to pass a valid middleware name, I think that in some cases it might not play well with remapping expansions. I don't see a problem particularly with DDS based `rmw` implementations, but if others need something more than just adding a prefix a suffix (e.g.: replacing the separator '/' with other specific separator), it might create problems.
Moreover, the idea of `no_mangle` is interacting with native applications of the underlying middleware, which probably don't have somthing similar to remapping and expansion, so it might make sense to ignore it.
I don't have arguments in favor of applying remapping and expansion when using `no_mangle=true`.

----

About adding a function that do both remapping and expansion in `rcl`, I think that's the correct path to go. I don't see an use case when you want to apply one and not the other to a topic/service name. Doing one and not the other just seems wrong.

Reply:
Maybe we could chat in person about this? I don't think we're talking about all the same things. For me (I may have the wrong idea), `no_mangle` may affect the output (as in do not demangle topic names in the returned information) or it may mean that the input should be used verbatim and not mangled (as in I'm passing a middleware topic name that should be used as-is, and not a ROS topic). In the latter case, I think that expanding the input may not be desired, but I could be wrong.

Reply:
My understanding is that this is indeed the latter case: if `no_mangle==true`, then the input into this `get_info_by_topic()` function and subsequently the `expand_topic_or_service_name()` function is to be treated as a middleware topic name.

-----------------------

> I don't have arguments in favor of applying remapping and expansion when using `no_mangle=true`.

> In the latter case, I think that expanding the input may not be desired, but I could be wrong.

Then it sounds like `no_mangle==true` does imply that the input `topic_name` parameter will need to be fully qualified? (Though we've been removing statements about that from the docstrings.)

Reply:
@ivanpauno summarized our discussion here:

https://github.com/ros2/ros2/issues/862#issuecomment-579464354

Basically we want all of the graph query functions (including this one) to do both expansion (`foo` becomes `/node_ns/foo`) and remapping, by default. And then we want them all to have an option to skip both expansion and remapping.

Also, any functions that return topic or type names should be able to avoid _de_-mangling of those names.

Currently this implementation does expansion but not remapping (as far as I can tell), and it has an option `no_mangle` to make the query assume the input topic should be considered as is, without any additional prefixes or other mangling. It does not, however, have a `no_demangle`, which some of the other functions do have.

This `no_mangle` option does not prevent the topic name expansion here, which is a problem in my opinion, what do you think @ivanpauno?

----

As I see it we need to do these things to get this merged:

- add remapping of the input topic name to this function
- avoid expansion and remapping when `no_mangle` is specified

Later we should make a set of pull requests which add a `no_demangle` option to these query functions. This will affect all the layers unfortunately.

Reply:
We should also make sure the `rclpy` version of this matches.

Reply:
> Basically we want all of the graph query functions (including this one) to do both expansion (foo becomes /node_ns/foo) and remapping, by default. And then we want them all to have an option to skip both expansion and remapping.

I completly agree with this.
I think it should be done in `rcl`, repeating the code in `rclcpp` and `rclpy` doesn't make much sense.

> Also, any functions that return topic or type names should be able to avoid de-mangling of those names.

Currently, the topic types are being demangled or not based on `no_mangle` argument.
Considering the use cases mentioned in the issue, I think that's fine.
We could have a better name, like `use_ros_conventions`, default to true:
- `true` means both input and output are ros names.
- `false` means both input and output are dds names (input shouldn't be expanded/remapped/mangled, output shouldn't be demangled).

> This no_mangle option does not prevent the topic name expansion here, which is a problem in my opinion, what do you think @ivanpauno?

Yeah, as we talked, I think it's a problem.

----

We can left this as-is for now, and then correct all the places where we have this error together (publisher/subscription creation is applying remapping and expansion when ros_conventions qos is false , publisher/subscription count is not doing remapping, etc). I don't feel strongly about which path to take.

Reply:
> I think it should be done in rcl, repeating the code in rclcpp and rclpy doesn't make much sense.

In principle yes, but actually I think the sub-nodes feature in rclcpp prevent this for now. It needs to be more carefully considered I think. I would postpone that to another pull request.

----

> Currently, the topic types are being demangled or not based on no_mangle argument.
Considering the use cases mentioned in the issue, I think that's fine.
We could have a better name, like use_ros_conventions, default to true: ...

I think having just one option can be ok, though it does make the odd cases impossible, because there's no way to manually do mangling/demangling.

If we want to go the route of having one option for both behaviors, then we need to change the `no_mangle` name, as it doesn't cover the existing behavior correctly, imo.

----

> We can left this as-is for now, and then correct all the places where we have this error together (publisher/subscription creation is applying remapping and expansion when ros_conventions qos is false , publisher/subscription count is not doing remapping, etc). I don't feel strongly about which path to take.

I don't think the two points I highlighted (add remapping and avoid expansion/remapping when `no_mangle` is true) are very difficult so I'd prefer those to be done now. Everything else can be done in follow up.